tdrs-backend/db-upgrade/cloud-foundry-db-upgrade.md

```diff
@@ -0,0 +1,204 @@
+# Cloud Foundry, Cloud.gov AWS RDS Database Upgrade
+The process below provides a guide to roll our backend applications over to a new RDS version and instance. The entire process can take several hours and does involve downtime for the environment which you are upgrading. Be sure to take those factors into account when commencing the process.
+
+## Process
+
+### 1. Open an SSH tunnel to the service
+To execute commands on the RDS instance we can open an SSH tunnel to the service and run all our commands from our local machine. Keep this tunnel open in a separate terminal window until this process is complete!
+
+```
+cf connect-to-service --no-client <APP_NAME_THAT_IS_BOUND_TO_RDS> <RDS_SERVICE_NAME>
+```
+
+You should see output similar to:
+
+```
+Finding the service instance details...
+Setting up SSH tunnel...
+SSH tunnel created.
+Skipping call to client CLI. Connection information:
+
+Host: localhost
+Port: 63634
+Username: <REDACTED>
+Password: <REDACTED>
+Name: <REDACTED>
+
+Leave this terminal open while you want to use the SSH tunnel. Press Control-C to stop.
+```
+
+### 2. Create a backup of the database(s) in the RDS instance
+In a separate terminal from your SSH tunnel terminal, generate the `pg_dump` files.
+Note: the HOST, PORT, DB_USER, and PASSWORD are the values you received from the output of the SSH tunnel. The DB_NAME parameter is the name of the DB you want to export, e.g `tdp_db_raft`. You will need to run this command for each DB in the instance.
+
+```
+pg_dump -h <HOST> -p <PORT> -d <DB_NAME> -U <DB_USER> -F c --no-acl --no-owner -f <FILE_NAME>.pg
+```
+
+After the command finishes, you should see <FILE_NAME>.pg in your current working directory.
+
+### 3. Update Terraform to create a new RDS instance
+Follow the instructions in the `terraform/README.md` to get Terraform configured. Modify the `main.tf` file in the `terraform/<ENV>` to include a new RDS instance. E.g if you were updating `prod` to version 15.x you would add the following code to the `main.tf` file. We are **NOT** removing the existing `resource "cloudfoundry_service_instance" "database"` from the `main.tf` file. Note that the resource name (i.e. `new-database`) and the `name` of the new RDS instance are not the same as the original resource name and RDS name. This is on purpose and we will remedy this in later steps.
+
+```yaml
+resource "cloudfoundry_service_instance" "new-database" {
+  name             = "tdp-db-prod-new"
+  space            = data.cloudfoundry_space.space.id
+  service_plan     = data.cloudfoundry_service.rds.service_plans["medium-gp-psql"]
+  json_params      = "{\"version\": \"15\", \"storage_type\": \"gp3\", \"storage\": 500}"
+  recursive_delete = true
+  timeouts {
+    create = "60m"
```

Comment:
these timeouts seem extensively high! for example, which if our create queries take 60 m?

Reply:
Just want to clarify that these timeouts are strictly for the terraform resource and not the database (and database queries) itself. Having these timeouts ensures that terraform monitors the infrastructure change until it is complete as opposed to timing out. In which case, the developer would have to query the state to see if the infrastructure has finished updating/creating/deleting. As another reference point, it took ~30 min just to delete an old elastic search service which did not have a lot of data in it and was a small service plan.